shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,279 @@
+package ssz
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// ErrNotMarakleRoot will be returned when a cache object is not a marakle root
+	ErrNotMarakleRoot = errors.New("object is not a markle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+
+	// Metrics
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_miss",
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "powchain_hash_cache_size",
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// markleRoot specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MarkleRoot []byte
+}
+
+// hashKeyFn takes the  representation as the key for a hashInfo.
+func hashKeyFn(obj interface{}) (string, error) {
+	mRoot, ok := obj.(*root)
+	if !ok {
+		return "", ErrNotMarakleRoot
+	}
+
+	return mRoot.Hash.Hex(), nil
+}
+
+// newHashCache creates a new hash cache for storing/accessing root hashes from
+// memory.
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: cache.NewFIFO(hashKeyFn),
+	}
+}
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *cache.FIFO
+}
+
+// RootByHash fetches Root by its hash. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByHash(h common.Hash) (bool, *root, error) {
+
+	obj, exists, err := b.hashCache.GetByKey(h.Hex())
+	if err != nil {
+		return false, nil, err
+	}
+
+	if exists {
+		hashCacheHit.Inc()
+	} else {
+		hashCacheMiss.Inc()
+		return false, nil, nil
+	}
+
+	hInfo, ok := obj.(*root)
+	if !ok {
+		return false, nil, ErrNotMarakleRoot
+	}
+
+	return true, hInfo, nil
+}
+
+// AddRetrieveTrieRoot adds a trie root to the cache. This method also trims the
+// least recently added root info if the cache size has reached the max cache
+// size limit.
+func (b *hashCacheS) AddRetrieveTrieRoot(val interface{}) ([32]byte, error) {
```

Comment:
The naming is odd, `AddRetrieveTrieRoot` signifies two opposite actions by this method.

Reply:
it retrieves from cache if this object is there and calculate the root and add it to cache when its missing

Reply:
Sure but add and retrieve mean two very different things, when I read the method I do not have a good idea what it does because the naming is confusing.

Reply:
got you. I hope I got the naming better this time